test/libsolutil/DominatorTest.cpp

```diff
@@ -0,0 +1,876 @@
+			return labels
+				| ranges::views::transform([&](TestVertexLabel const& label) {
+					soltestAssert(verticesIdMap.count(label));
+					return verticesIdMap.at(label);
+				})
+				| ranges::to<std::vector<TestVertexId>>;
+		};
+
+		return toVertexMapById(_vertexMapByLabel | ranges::views::transform([&](auto const& pair) -> std::pair<TestVertexLabel, std::vector<TestVertexId>> {
+			return {pair.first, labelVectorToIdVector(pair.second)};
+		}) | ranges::to<std::map<TestVertexLabel, std::vector<TestVertexId>>>);
+	}
+
+	TestVertex const* entry = nullptr;
+	std::vector<TestVertex> vertices;
+	// Reverse map from vertices labels to ids
+	std::map<TestVertexLabel, TestVertexId> verticesIdMap;
+	std::map<TestVertexId, TestVertexId> expectedIdom;
+	std::map<TestVertexId, size_t> expectedDFSIndices;
+	std::map<TestVertexId, std::vector<TestVertexId>> expectedDominatorTree;
+};
+
+BOOST_AUTO_TEST_SUITE(Dominators)
+
+BOOST_AUTO_TEST_CASE(immediate_dominator_1)
+{
+	//            A
+	//            │
+	//            ▼
+	//        ┌───B
+	//        │   │
+	//        ▼   │
+	//        C ──┼───┐
+	//        │   │   │
+	//        ▼   │   ▼
+	//        D◄──┘   G
+	//        │       │
+	//        ▼       ▼
+	//        E       H
+	//        │       │
+	//        └──►F◄──┘
+	DominatorFinderTest test(
+		{"A", "B", "C", "D", "E", "F", "G", "H"}, // Vertices
+		{ // Edges
+			Edge("A", "B"),
+			Edge("B", "C"),
+			Edge("B", "D"),
+			Edge("C", "D"),
+			Edge("C", "G"),
+			Edge("D", "E"),
+			Edge("E", "F"),
+			Edge("G", "H"),
+			Edge("H", "F")
+		},
+		{ // Immediate dominators
+			{"A", "A"},
+			{"B", "A"},
+			{"C", "B"},
+			{"D", "B"},
+			{"E", "D"},
+			{"F", "B"},
+			{"G", "C"},
+			{"H", "G"}
+		},
+		{ // DFS indices
+			{"A", 0},
+			{"B", 1},
+			{"C", 2},
+			{"D", 3},
+			{"E", 4},
+			{"F", 5},
+			{"G", 6},
+			{"H", 7}
+		},
+		{ // Dominator tree
+			{"A", {"B"}},
+			{"B", {"C", "D", "F"}},
+			{"C", {"G"}},
+			{"D", {"E"}},
+			{"G", {"H"}}
+		}
+	);
+
+	TestDominatorFinder dominatorFinder(*test.entry);
+	BOOST_TEST(dominatorFinder.idom() == test.expectedIdom);
+	BOOST_TEST(dominatorFinder.dfsIndexById() == test.expectedDFSIndices);
+	BOOST_TEST(dominatorFinder.dominatorTree() == test.expectedDominatorTree);
+}
+
+BOOST_AUTO_TEST_CASE(immediate_dominator_2)
+{
+	//    ┌────►A──────┐
+	//    │     │      ▼
+	//    │ B◄──┘   ┌──D──┐
+	//    │ │       │     │
+	//    │ ▼       ▼     ▼
+	//    └─C◄───┐  E     F
+	//      │    │  │     │
+	//      └───►G◄─┴─────┘
+	DominatorFinderTest test(
+		{"A", "B", "C", "D", "E", "F", "G"},
+		{
+			Edge("A", "B"),
+			Edge("B", "C"),
+			Edge("C", "G"),
+			Edge("C", "A"),
+			Edge("A", "D"),
+			Edge("D", "E"),
+			Edge("D", "F"),
+			Edge("E", "G"),
+			Edge("F", "G"),
+			Edge("G", "C")
+		},
+		{
+			{"A", "A"},
+			{"B", "A"},
+			{"C", "A"},
+			{"D", "A"},
+			{"E", "D"},
+			{"F", "D"},
+			{"G", "A"}
```

Comment:
I just scrolled through the tests.Are these immediate dominators really correct? Why is it `A` for `A` and not `C`? I guess it's somehow a special case or something like this, right? Maybe the others are also somehow special cases. However, I was also wondering with `C`, why is it `A` and not `B` or `G`?  Also `G` why is it `A` and not `E` or `F` or `C`?

Reply:
Maybe I should just read the paper ;)

Reply:
Indeed, there is a mistake in the immediate dominator set due to a misleading representation of the result.
`A` does not have any immediate dominator, as no node dominates `A` besides `A` itself. I believe this was the reason of your confusion with the other values.

I should have used an `std::optional` to distinguish the special case of the entry vertex (i.e. vertex `A` in this case) in the immediate dominator vector.
The `m_immediateDominators` vector associates the `dfsIndex` of a vertex with the `dfsIndex` of its immediate dominator.
For example, if vertex `w` immediate dominates vertex `v`, it is represented as: `m_immediateDominators[w] = v`.

The issue in my implementation is that `DfsIndex` is an `unsigned integer`, and the entry vertex will always have minimal `DfsIndex` possible, i.e. `0`.
Thus, I should have used a default value other than zero to represent the entry vertex, for example: `m_immediateDominators['A'] = std::nullopt`.

However, the other values in the example are correct. To clarify:

The immediate dominator of a vertex `v` is the vertex `w` iff:
- `w` strictly dominates `v`
- `w` does not strictly dominate any other vertex that strictly dominates `v`

A vertex `w` strictly dominates `v` if `w` dominates `v` and `w != v`. By definition, every vertex dominates itself.
Also, note that in my implementation, the self-dominance relationship is implicit.

In that example, `A` dominates `B` because any path from the entry vertex to `B` must pass through `A`.
However, `B` does not dominate `C` because there are other paths that lead to `C` without going through `B`,
for instance, `A -> D -> E -> G -> C`.
By similar reasoning, `G` does not dominate `C` and neither `E`, `F`, nor `D` dominates `C`, but `A` does.
To be an immediate dominator, a vertex must strictly dominate the other.

The dominators of `C` are `{A, C}`, and by definition of strict dominance, `C` cannot be the immediate dominator of itself. Therefore, `A` is the closest dominator of `C`, thus its immediate dominator.

Also, `E`, `F` and `C` do not dominates `G`. In this example, the only vertex that you necessarily need to pass to reach `G` from the entry is `A`.

Reply:
> Maybe I should just read the paper ;)

Haha, I think the mistake with the entry vertex made the entire example confusing.